---
title: "Single Cell RNA-Seq Analysis"
author: "Pallavi Gaur"
date: '2022-04-28'
output:
  html_document:
    df_print: paged
---
This is a tutorial for running single-cell/nucleus RNA-seq analysis. We will be using subsetted data from 2,000 individual nuclei from Cain et al. (Nat Neuro 2023). We will start with the pre -processing and walking through the steps like normalization, scaling , PCA, UMAP projection, we will be clustering the cells , trying to identify the major cell types and looking into the markers and DEGs in the clusters.

```{r}
packages_to_install=c("ggplot2","dplyr")
install.packages(setdiff(packages_to_install, rownames(installed.packages())))
```

```{r}
BiocManager::install("Seurat")
```

### Libraries used
```{r,warning=FALSE}
require(Seurat)
require(ggplot2)
require(dplyr)
```

## Load the counts table
```{r}
# load the data
counts=readRDS("Session5_snrnaseq_data.rds")
```

## Create a Seurat object from the counts table - this object will contain all of our data and analysis results
```{r}
data1=CreateSeuratObject(counts)
```

##Standard pre-processing workflow
The steps below encompass the standard pre-processing workflow for sc/snRNA-seq data in Seurat. These represent the selection and filtration of cells based on QC metrics, data normalization and scaling, and the detection of highly variable features (genes).

### QC and selecting cells for further analysis
Seurat allows you to easily explore QC metrics and filter cells based on any user-defined criteria. A few QC metrics commonly used by the community include:
1) The number of unique genes detected in each cell. Low-quality cells or empty droplets will often have very few genes, whereas  doublets or multiplets may exhibit an aberrantly high gene count.
2) Similarly, the total number of molecules detected within a cell (correlates strongly with unique genes)
3) The percentage of reads that map to the mitochondrial genome. Low-quality / dying cells often exhibit extensive mitochondrial contamination. We calculate mitochondrial QC metrics with the PercentageFeatureSet() function, which calculates the percentage of counts originating from a set of features. We use the set of all genes starting with MT- as a set of mitochondrial genes

```{r}
data1[["percent.mt"]] <- PercentageFeatureSet(data1, pattern = "^MT-")
```

### Visualize QC metrics violin plots
Here we visualize QC metrics and use these to filter cells.
We filter out cells that have unique features (genes) over 4000 or less than 500
We filter out cells that have >5% mitochondrial counts

```{r}
VlnPlot(data1, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```
#### Filtering the data
```{r}
data1_filt <- subset(data1, subset = nFeature_RNA <=4000 &  nFeature_RNA >=500 & percent.mt < 5)
data1_filt
```

### Normalizing the data
After removing unwanted cells from the dataset, the next step is to normalize the data. By default, we employ a global-scaling normalization method "LogNormalize" that divides the counts for each gene in each cell by the total counts for that cell, multiplies that value by 10,000, and then log-transforms the result. Normalized values are stored in data1[["RNA"]]@data.
```{r}
data1_filt <- NormalizeData(data1_filt, normalization.method = "LogNormalize", scale.factor = 10000)
```
### Identification of highly variable features (feature selection)
We next calculate a subset of features that exhibit high cell-to-cell variation in the dataset (i.e, they are highly expressed in some cells, and lowly expressed in others). Focusing on these genes in downstream analysis helps to highlight biological signal in single-cell datasets, and to group cells into clusters.

The function FindVariableFeatures() directly models the mean-variance relationship inherent in single-cell data (taking the top 2,000 most variable features by default). These will be used in downstream analysis, like PCA and clustering.
```{r}
data1_filt <- FindVariableFeatures(data1_filt, selection.method = "vst", nfeatures = 2000)
# Look at the 10 most highly variable genes
top20 <- head(VariableFeatures(data1_filt), 20)
# plot variable features with and without labels
plot1 <- VariableFeaturePlot(data1_filt)
plot1 <- LabelPoints(plot = plot1, points = top20, repel = TRUE)
plot1
```

### Scaling the data
Next, we apply a linear transformation ('scaling') that is a standard pre-processing step prior to dimensional reduction techniques like PCA. This is to put all the highly variable genes on a level playing field, so that those with the highest expression do not overwhelm genes with lower expression.
In Seurat, the ScaleData() function performs a "z-score" transformation:
It shifts the expression of each gene, so that the mean expression across cells is 0
It scales the expression of each gene, so that the variance across cells is 1
This step gives equal weight in downstream analyses, so that highly-expressed genes do not dominate
The results of this are stored in dataname[["RNA"]]@scale.data
```{r}
all.genes <- rownames(data1_filt)
data1_scale <- ScaleData(data1_filt, features = all.genes)
data1_scale
```

The ScaleData() function can also be used to remove unwanted sources of variation from a single-cell dataset. For example, we could 'regress out' heterogeneity associated with mitochondrial contamination. 
```{r}
#data1_scale <- ScaleData(pbmc, vars.to.regress = "percent.mt")
```

###Perform linear dimensional reduction
Next we perform PCA on the scaled data, and select the number of PCs to keep
```{r}
data1_filt <- RunPCA(data1_scale,features = VariableFeatures(object = data1_filt))
```
Seurat provides several useful ways of visualizing both cells and features that define the PCA, including VizDimReduction(), DimPlot(), and DimHeatmap()

```{r}
DimHeatmap(data1_filt, dims = 1:2, cells = 500, balanced = TRUE)
```

### Select the number of PCs to keep, based on the elbow plot
'Elbow plot': a ranking of principal components based on the percentage of variance explained by each one (ElbowPlot() function). In this example, we can observe a (weak) 'elbow' around PC17, suggesting that the majority of true signal is captured in the first 17 PCs.

```{r}
ElbowPlot(data1_filt)
```

### Cluster the cells
Seurat applies a graph-based clustering approach, building upon initial strategies in (Macosko et al. 2016). 
The distance metric for clustering is calculated as follows: 1) Calculate the Euclidean distance in PCA space (using the number of PCs we have selected), 2) Construct a k-Nearest Neighbors graph, 3) Find the overlap between any two cells based on how many nearest neighbors they share (Jaccard overlap). The inverse of this overlap is the distance between any two cells.
The algorithm to group cells together is based on a modularity optimization technique such as the Louvain algorithm (default) or SLM [SLM, Blondel et al., Journal of Statistical Mechanics], to iteratively group cells together, with the goal of optimizing the standard modularity function. The FindClusters() function implements this procedure, and contains a resolution parameter that sets the 'granularity' of the downstream clustering, with increased values leading to a greater number of clusters. 
We find that setting the resolution parameter between 0.3-1.2 typically returns good results for single-cell datasets of around 3K cells. Optimal resolution often increases for larger datasets, and this resolution parameter can be explored systematically.
```{r}
data1_filt <- FindNeighbors(data1_filt, dims = 1:17)
data1_filt <- FindClusters(data1_filt, resolution = 0.5)
```

###Visualizing cells in a 2-dimensional space
Seurat offers several dimensional reduction approaches for visualization of the single-cell data, such as PCA, tSNE, and UMAP. The goal of these algorithms is to learn the underlying manifold of the data in order to place similar cells together in low-dimensional space. 
Cells within the graph-based clusters determined above should co-localize on these dimension reduction plots, although cluster boundaries may not show 100% separation int he visualization. As input to the UMAP and tSNE, we suggest using the same PCs as input to the clustering analysis.
```{r}
# If you haven't installed UMAP, you can do so via reticulate::py_install(packages =
# 'umap-learn')
data1_filt <- RunTSNE(data1_filt, dims = 1:15)
data1_filt <- RunUMAP(data1_filt, dims = 1:15)
```

###Explore different visualizations of the data
```{r}
DimPlot(data1_filt,reduction="pca",group.by="seurat_clusters",label=T)
DimPlot(data1_filt,reduction="tsne",group.by="seurat_clusters",label=T)
DimPlot(data1_filt,reduction="umap",group.by="seurat_clusters",label=T)
```

### Biological interpretation of clusters
Next, we want to examine whether these clusters we have found through Seurat show specific expression of genes known to distinguish different cell types in the brain. If this clustering is biologically relevant, we would hope that some of these clusters show restricted expression of certain genes. In the human cortex, certain genes are known to "mark" certain cell types:
All neurons (RBFOX3+)
Excitatory neurons (SLC17A7+)
Inhibitory neurons (GAD2+, SLC32A1+)
Astrocytes (AQP4+, FGFR3+, ALDH1L1+)
Oligodendrocytes (MOG+)
Oligodendrocyte precursor cells (PDGFRA+)
Microglia (AIF1+, C1QA+)
Endothelial cells (CLDN5+)

###Visualize the expression of specific genes using a "Dotplot"
```{r}
features<-c("RBFOX3","SLC17A7","GAD2","SLC32A1",
            "AQP4","FGFR3", "ALDH1L1","MOG","PDGFRA",
            "AIF1","C1QA","CLDN5")
DotPlot(data1_filt,
            features  = features,
            cols = c("#70e1f5","#B00B1E"), dot.scale = 8,scale.max = 100,scale.min = 0)+
theme(axis.text.x = element_text(size = 12),
      axis.text.y = element_text(size = 12))+RotatedAxis()
```

###Visualize the expression of specific genes on the UMAP plot, showing expression in each cell
```{r fig.width = 19, fig.height=14}
features<-c("RBFOX3","SLC17A7","GAD2",
            "AQP4","MOG","PDGFRA",
            "AIF1","CLDN5")
FeaturePlot(data1_filt, 
            reduction = "umap", 
            features =features, 
            order = TRUE,
            min.cutoff = 'q10')
```
### From unsupervised to supervised analysis: finding differential genes between the clusters we have identified
To better characterize the clusters, Seurat can help you find markers that distinguish clusters from each other via differential expression. It can identify genes that are more highly expressed in a single cluster (specified in ident.1), compared to all other cells, using a non-parametric Mann-Whitney test.
```{r}
# find markers for every cluster compared to all remaining cells, report only the positive
# ones
cluster0_markers <- FindMarkers(data1_filt, ident.1=0, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
head(cluster0_markers)
```

### Visualize these "new" markers using Dotplots, Violin plots, and on the UMAP
Seurat provides several tools for visualizing marker expression:
1) DotPlot() - as seen above, this shows the fraction of cells within each cluster in which a gene is detected, as well as its mean detection value.
2) FeaturePlot() - as seen above, this visualizes the expression of a feature in every cell on the 2-D reduced dimension visualization
3) VlnPlot()  - sshows expression distributions within each cluster. Violin plots are histograms that are plotted vertically and reflected along the y-axis
```{r}
features_to_plot = c("SLC1A2","GPC5")
DotPlot(data1_filt, features=features_to_plot)
FeaturePlot(data1_filt, reduction = "umap", features=features_to_plot)
VlnPlot(data1_filt, features=features_to_plot)
```

#### Visualization using a heatmap
Here, every column is a cell (ordered by cluster), and every row represents a gene
```{r}
DoHeatmap(data1_filt, features = features_to_plot) + NoLegend()
```

### Exploration
1) Try changing the resolution parameter to see how the number of clusters changes
2) Find markers for a different cluster and visualize the expression of these markers





